File: CompanionLib/Utility/FBiOSTargetProvider.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBiOSTargetProvider.h"

#import <FBDeviceControl/FBDeviceControl.h>
#import <FBSimulatorControl/FBSimulatorControl.h>
#import <XCTestBootstrap/XCTestBootstrap.h>

#import "FBIDBError.h"

@implementation FBiOSTargetProvider

#pragma mark Public

+ (FBFuture<id<FBiOSTarget>> *)targetWithUDID:(NSString *)udid targetSets:(NSArray<id<FBiOSTargetSet>> *)targetSets warmUp:(BOOL)warmUp logger:(id<FBControlCoreLogger>)logger
{
  NSError *error = nil;
  if ([udid.lowercaseString isEqualToString:@"only"]) {
    id<FBiOSTarget> target = [self fetchSoleTargetForTargetSets:targetSets logger:logger error:&error];
    if (!target) {
      return [FBFuture futureWithError:error];
    }
    return [FBFuture futureWithResult:target];
  }
  id<FBiOSTarget> target = [self fetchTargetWithUDID:udid targetSets:targetSets logger:logger error:&error];
  if (!target) {
    return [FBFuture futureWithError:error];
  }
  if (!warmUp) {
    return [FBFuture futureWithResult:target];
  }
  if (target.state != FBiOSTargetStateBooted) {
    return [FBFuture futureWithResult:target];
  }
  id<FBSimulatorLifecycleCommands> lifecycle = (id<FBSimulatorLifecycleCommands>) target;
  if (![lifecycle conformsToProtocol:@protocol(FBSimulatorLifecycleCommands)]) {
    return [FBFuture futureWithResult:target];;
  }
  
  if (FBXcodeConfiguration.isXcode12_5OrGreater) {
    return [FBFuture futureWithResult:target];
  }
  
  return [[lifecycle
    connectToBridge]
    mapReplace:target];
}

#pragma mark Private

+ (id<FBiOSTarget>)fetchTargetWithUDID:(NSString *)udid targetSets:(NSArray<id<FBiOSTargetSet>> *)targetSets logger:(id<FBControlCoreLogger>)logger error:(NSError **)error
{
  // Get a mac device if one was requested
  if ([udid.lowercaseString isEqualToString:@"mac"]) {
    return [[FBMacDevice alloc] initWithLogger:logger];
  }
  // Otherwise query the input target sets
  for (id<FBiOSTargetSet> targetSet in targetSets) {
    id<FBiOSTargetInfo> targetInfo = [targetSet targetWithUDID:udid];
    if (!targetInfo) {
      continue;
    }
    if (![targetInfo conformsToProtocol:@protocol(FBiOSTarget)]) {
      return [[FBDeviceControlError
        describeFormat:@"UDID %@ exists, but the target is not usable %@", udid, targetInfo]
        fail:error];
    }
    return (id<FBiOSTarget>) targetInfo;
  }

  return [[FBIDBError
    describeFormat:@"%@ could not be resolved to any target in %@", udid, targetSets]
    fail:error];
}

+ (id<FBiOSTarget>)fetchSoleTargetForTargetSets:(NSArray<id<FBiOSTargetSet>> *)targetSets logger:(id<FBControlCoreLogger>)logger error:(NSError **)error
{
  NSMutableArray<id<FBiOSTarget>> *targets = NSMutableArray.array;
  for (id<FBiOSTargetSet> targetSet in targetSets) {
    [targets addObjectsFromArray:(NSArray<id<FBiOSTarget>> *)targetSet.allTargetInfos];
  }
  if (targets.count > 1) {
    return [[FBIDBError
      describeFormat:@"Cannot get a sole target when multiple found %@", [FBCollectionInformation oneLineDescriptionFromArray:targets]]
      fail:error];
  }
  if (targets.count == 0) {
    return [[FBIDBError
      describeFormat:@"Cannot get a sole target when none were found in target sets %@", [FBCollectionInformation oneLineDescriptionFromArray:targetSets]]
      fail:error];
  }
  return targets.firstObject;
}

@end

File: CompanionLib/Utility/FBiOSTargetProvider.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import FBControlCore
@_implementationOnly import FBDeviceControl
import FBSimulatorControl
import Foundation
import XCTestBootstrap

/// The ways target resolution can fail, as data rather than assembled strings.
public enum FBiOSTargetProviderError: Error {
  case targetNotUsable(udid: String, targetDescription: String)
  case targetNotFound(udid: String, targetSetsDescription: String)
  case multipleTargets(targetsDescription: String)
  case noTargets(targetSetsDescription: String)
  case multipleBootedTargets(targetsDescription: String)
  case noBootedTargets(targetSetsDescription: String)
}

extension FBiOSTargetProviderError: LocalizedError {
  public var errorDescription: String? {
    switch self {
    case let .targetNotUsable(udid, targetDescription):
      return "\(udid) exists, but the target is not usable \(targetDescription)"
    case let .targetNotFound(udid, targetSetsDescription):
      return "\(udid) could not be resolved to any target in \(targetSetsDescription)"
    case let .multipleTargets(targetsDescription):
      return "Cannot get a sole target when multiple found \(targetsDescription)"
    case let .noTargets(targetSetsDescription):
      return "Cannot get a sole target when none were found in target sets \(targetSetsDescription)"
    case let .multipleBootedTargets(targetsDescription):
      return "Cannot get a sole booted target when multiple are booted \(targetsDescription)"
    case let .noBootedTargets(targetSetsDescription):
      return "Cannot get a sole booted target when none are booted in target sets \(targetSetsDescription)"
    }
  }
}

public final class FBiOSTargetProvider {

  public static func target(withUDID udid: String, targetSets: [FBiOSTargetSet], warmUp: Bool, logger: FBControlCoreLogger) throws -> FBiOSTarget {
    switch udid.lowercased() {
    case "only":
      return try fetchSoleTarget(forTargetSets: targetSets, logger: logger)
    case "booted":
      return try fetchSoleBootedTarget(forTargetSets: targetSets, logger: logger)
    default:
      return try fetchTarget(withUDID: udid, targetSets: targetSets, logger: logger)
    }
  }

  // MARK: - Private

  private static func fetchTarget(withUDID udid: String, targetSets: [FBiOSTargetSet], logger: FBControlCoreLogger) throws -> FBiOSTarget {
    if udid.lowercased() == "mac" {
      return FBMacDevice(logger: logger)
    }
    for targetSet in targetSets {
      guard let targetInfo = targetSet.target(withUDID: udid) else {
        continue
      }
      guard let target = targetInfo as? FBiOSTarget else {
        throw FBiOSTargetProviderError.targetNotUsable(udid: udid, targetDescription: String(describing: targetInfo))
      }
      return target
    }

    throw FBiOSTargetProviderError.targetNotFound(udid: udid, targetSetsDescription: String(describing: targetSets))
  }

  private static func fetchSoleTarget(forTargetSets targetSets: [FBiOSTargetSet], logger: FBControlCoreLogger) throws -> FBiOSTarget {
    var targets: [FBiOSTarget] = []
    for targetSet in targetSets {
      for info in targetSet.allTargetInfos {
        if let target = info as? FBiOSTarget {
          targets.append(target)
        }
      }
    }
    if targets.count > 1 {
      throw FBiOSTargetProviderError.multipleTargets(targetsDescription: FBCollectionInformation.oneLineDescription(from: targets))
    }
    guard let target = targets.first else {
      throw FBiOSTargetProviderError.noTargets(targetSetsDescription: FBCollectionInformation.oneLineDescription(from: targetSets))
    }
    return target
  }

  private static func fetchSoleBootedTarget(forTargetSets targetSets: [FBiOSTargetSet], logger: FBControlCoreLogger) throws -> FBiOSTarget {
    var bootedTargets: [FBiOSTarget] = []
    for targetSet in targetSets {
      for info in targetSet.allTargetInfos {
        guard let target = info as? FBiOSTarget, target.state == .booted else {
          continue
        }
        bootedTargets.append(target)
      }
    }
    if bootedTargets.count > 1 {
      throw FBiOSTargetProviderError.multipleBootedTargets(targetsDescription: FBCollectionInformation.oneLineDescription(from: bootedTargets))
    }
    guard let target = bootedTargets.first else {
      throw FBiOSTargetProviderError.noBootedTargets(targetSetsDescription: FBCollectionInformation.oneLineDescription(from: targetSets))
    }
    return target
  }
}
